source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) 2020 André Perez Maselco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_inline_function.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationInlineFunction::TransformationInlineFunction(
+    const spvtools::fuzz::protobufs::TransformationInlineFunction& message)
+    : message_(message) {}
+
+TransformationInlineFunction::TransformationInlineFunction(
+    const std::map<uint32_t, uint32_t>& result_id_map,
+    uint32_t function_call_id) {
+  *message_.mutable_result_id_map() =
+      google::protobuf::Map<google::protobuf::uint32, google::protobuf::uint32>(
+          result_id_map.begin(), result_id_map.end());
+  message_.set_function_call_id(function_call_id);
+}
+
+bool TransformationInlineFunction::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // The values in the |message_.result_id_map| must be all fresh and all
+  // distinct.
+  std::set<uint32_t> ids_used_by_this_transformation;
+  for (auto& pair : message_.result_id_map()) {
+    if (!CheckIdIsFreshAndNotUsedByThisTransformation(
+            pair.second, ir_context, &ids_used_by_this_transformation)) {
+      return false;
+    }
+  }
+
+  // |function_call_instruction| must be defined, must be an OpFunctionCall
+  // instruction, must be the penultimate instruction in its block and its block
+  // termination instruction must be an OpBranch.
+  auto function_call_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.function_call_id());
+  auto function_call_instruction_block =
+      ir_context->get_instr_block(function_call_instruction);
```

Comment:
I don't think so, because I've already checked that `function_call_instruction` is not `nullptr` and it is an `SpvOpFunctionCall` instruction. Therefore, `function_call_instruction_block` can not be `nullptr` because an `SpvOpFunctionCall` instruction is only valid inside a block.

Reply:
`message_.function_call_id()` might not exist in the module. So `function_call_instruction = nullptr`. Then we call `ir_context->get_instr_block(nullptr);`. I believe this won't crash, but it is not ideal. I suggest you check that `function_call_instruction` is non-null immediately, like so:

```
auto function_call_instruction =
      ir_context->get_def_use_mgr()->GetDef(message_.function_call_id());
if (!function_call_instruction) {
  return false;
}
```

Just because this is clearer.

Reply:
But this is already done in the following `if` statement.

Reply:
Done.